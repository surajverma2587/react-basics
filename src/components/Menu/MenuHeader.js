import React from 'react';
import PropTypes from 'prop-types';
/* external library to render the stars for ratings */
/* https://www.npmjs.com/package/react-star-ratings */
import StarRatings from 'react-star-ratings';

const MenuHeader = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* replace the bottom return statement with the populated html template */
  return (
    <div>
      Menu Header Component
    </div>
  );
};

/* display name for the component */
MenuHeader.displayName = 'MenuHeader';

MenuHeader.propTypes = {
  /* add Proptypes here */
};

export default MenuHeader;