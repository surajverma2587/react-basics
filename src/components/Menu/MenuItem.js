import React from 'react';
import PropTypes from 'prop-types';

const MenuItem = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* validate the props and render the component IF valid ELSE null */
  /* replace the bottom return statement with the populated html template */
  return (
    <div>
      MenuItem Component
    </div>
  );
};

/* display name for the component */
MenuItem.displayName = 'MenuItem';

MenuItem.propTypes = {
  /* add Proptypes here */
};

export default MenuItem;