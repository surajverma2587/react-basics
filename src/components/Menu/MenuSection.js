import React from 'react';
import PropTypes from 'prop-types';
import MenuItem from './MenuItem';

const renderMenuItems = (/* menuItems */) => {
  /* map through the menuItems prop and return a list of MenuItem components */
};

const MenuSection = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* validate the props and render the component IF valid ELSE null */
  /* remember to call renderMenuItems() to render a list of MenuItem components */
  /* replace the bottom return statement with the populated html template */
  return (
    <div>
      MenuSection Component
    </div>
  );
};

/* display name for the component */
MenuSection.displayName = 'MenuSection';

MenuSection.propTypes = {
  /* add Proptypes here */
};

export default MenuSection;