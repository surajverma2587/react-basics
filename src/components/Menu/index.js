import React from 'react';
import PropTypes from 'prop-types';
/* external library to render the stars for ratings */
/* https://www.npmjs.com/package/react-star-ratings */
import StarRatings from 'react-star-ratings';
import MenuSection from './MenuSection';
import './styles.css'

const renderMenuSections = (/* menuSections */) => {
  /* map through the menuSections prop and return a list of MenuSection components */
};

const Menu = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* validate the props and render the component IF valid ELSE null */
  /* remember to call renderMenuSections() to render a list of MenuSection components */
  /* replace the bottom return statement with the populated html template */
};

export default Menu;