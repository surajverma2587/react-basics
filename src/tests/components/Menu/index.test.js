import React from 'react';
import { shallow } from 'enzyme';

import Menu from '../../../components/Menu';

const props = {
  id: 'M1',
  title: 'Menu title',
  address: 'Menu address',
  imgUrl: 'https://google.com',
  reviews: 100,
  rating: 2.5,
  menuSections: [{
    id: 'S1',
    title: 'Section title 1',
    description: 'Section description 1',
    menuItems: [
      {
        id: 'I1',
        title: 'Item title 1',
        description: 'Item description 1',
        price: 4
      },
      {
        id: 'I2',
        title: 'Item title 2',
        description: 'Item description 2',
        price: 5
      },
    ],
    onClickMenuItem: jest.fn(),
  }, {
    id: 'S2',
    title: 'Section title 2',
    description: 'Section description 2',
    menuItems: [
      {
        id: 'I3',
        title: 'Item title 3',
        description: 'Item description 3',
        price: 2
      },
      {
        id: 'I4',
        title: 'Item title 4',
        description: 'Item description 4',
        price: 8
      },
    ],
    onClickMenuItem: jest.fn(),
  }],
  onClickMenuItem: jest.fn()
};

describe('Menu Component', () => {
  describe('when props are valid', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<Menu {...props} />);
    });
    
    it('should render snapshot of the component', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should render the component', () => {
      const restaurant = wrapper.find('.restaurant');
      expect(restaurant.exists()).toBe(true);
    });

    it('should render the MenuHeader component', () => {
      const menuHeader = wrapper.find('MenuHeader');
      expect(menuHeader.length).toEqual(1);
      expect(menuHeader.prop('title')).toEqual(props.title);
      expect(menuHeader.prop('address')).toEqual(props.address);
      expect(menuHeader.prop('imgUrl')).toEqual(props.imgUrl);
      expect(menuHeader.prop('reviews')).toEqual(props.reviews);
      expect(menuHeader.prop('rating')).toEqual(props.rating);
    });
  
    it('should render the MenuSection components', () => {
      const menuSections = wrapper.find('MenuSection');
      expect(menuSections.length).toEqual(2);
      menuSections.forEach((menuSection, index) => {
        expect(menuSection.prop('id')).toEqual(props.menuSections[index].id);
        expect(menuSection.prop('title')).toEqual(props.menuSections[index].title);
        expect(menuSection.prop('description')).toEqual(props.menuSections[index].description);
        expect(menuSection.prop('menuItems')).toEqual(props.menuSections[index].menuItems);
      })
    });
  });

  describe('when props are invalid', () => {
    it('should not render the component when props are not present', () => {
      const wrapper = shallow(<Menu />);  
      expect(wrapper.find('.restaurant').exists()).toBe(false);
    });

    it('should not render the component when any prop is invalid', () => {
      const invalidProps = [
        { ...props, id: null },
        { ...props, title: null },
        { ...props, address: null },
        { ...props, menuSections: null },
        { ...props, menuSections: [] },
        { ...props, onClickMenuItem: null },
      ];
      invalidProps.forEach(invalidProp => {
        const wrapper = shallow(<Menu {...invalidProp} />);
        expect(wrapper.find('.restaurant').exists()).toBe(false);
      });
    });
  });
});
