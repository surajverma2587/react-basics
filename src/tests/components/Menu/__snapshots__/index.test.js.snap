// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Component when props are valid should render snapshot of the component 1`] = `
<div
  className="restaurant"
>
  <MenuHeader
    address="Menu address"
    imgUrl="https://google.com"
    rating={2.5}
    reviews={100}
    title="Menu title"
  />
  <div
    className="menu"
  >
    <MenuSection
      description="Section description 1"
      id="S1"
      key="S1"
      menuItems={
        Array [
          Object {
            "description": "Item description 1",
            "id": "I1",
            "price": 4,
            "title": "Item title 1",
          },
          Object {
            "description": "Item description 2",
            "id": "I2",
            "price": 5,
            "title": "Item title 2",
          },
        ]
      }
      onClickMenuItem={[MockFunction]}
      title="Section title 1"
    />
    <MenuSection
      description="Section description 2"
      id="S2"
      key="S2"
      menuItems={
        Array [
          Object {
            "description": "Item description 3",
            "id": "I3",
            "price": 2,
            "title": "Item title 3",
          },
          Object {
            "description": "Item description 4",
            "id": "I4",
            "price": 8,
            "title": "Item title 4",
          },
        ]
      }
      onClickMenuItem={[MockFunction]}
      title="Section title 2"
    />
  </div>
</div>
`;
